src/object/cli_obj.c

```diff
@@ -1616,126 +1988,60 @@ dc_obj_update(tse_task_t *task)
 		rc = -DER_NO_HDL;
 		goto out_task;
 	}
+	rc = obj_ptr2pm_ver(obj, &map_ver);
+	if (rc) {
+		obj_decref(obj);
+		goto out_task;
+	}
 
 	oca = daos_oclass_attr_find(obj->cob_md.omd_id);
 	if (oca->ca_resil == DAOS_RES_EC) {
 		rc = ec_obj_update_encode(task, obj->cob_md.omd_id, oca,
 					  &tgt_set);
-		if (rc != 0)
+		if (rc != 0) {
+			obj_decref(obj);
 			goto out_task;
+		}
 	}
 
 	obj_auxi = tse_task_stack_push(task, sizeof(*obj_auxi));
 	obj_auxi->opc = DAOS_OBJ_RPC_UPDATE;
+	obj_auxi->map_ver_req = map_ver;
+	obj_auxi->obj_task = task;
 	shard_task_list_init(obj_auxi);
 	rc = tse_task_register_comp_cb(task, obj_comp_cb, &obj,
 				       sizeof(obj));
 	if (rc != 0) {
-		/* NB: obj_comp_cb() will release refcount in other cases */
 		obj_decref(obj);
 		goto out_task;
 	}
 
-	rc = obj_ptr2pm_ver(obj, &map_ver);
-	if (rc)
-		goto out_task;
-
 	dkey_hash = obj_dkey2hash(args->dkey);
-	rc = obj_dkeyhash2update_grp(obj, dkey_hash, map_ver, &shard,
-				     &shards_cnt);
-	if (rc != 0)
-		goto out_task;
-	start_shard = shard;
-
-	D_DEBUG(DB_IO, "update "DF_OID" dkey %llu start %u cnt %u\n",
-		DP_OID(obj->cob_md.omd_id), (unsigned long long)dkey_hash,
-		shard, shards_cnt);
-	obj_auxi->fw_shard_tgts = &obj_auxi->fw_shard_tgts_inline;
-	obj_auxi->fw_cnt = &obj_auxi->fw_cnt_inline;
-	obj_auxi->fw_rdg_count = 1;
-	rc = obj_shards_2_fwtgts(obj, map_ver, tgt_set, &shard, &shards_cnt,
-				 obj_auxi->fw_shard_tgts, obj_auxi->fw_cnt);
-	if (rc != 0) {
-		D_ERROR("update "DF_OID", obj_shards_2_fwtgts failed %d.\n",
-			DP_OID(obj->cob_md.omd_id), rc);
+	rc = obj_req_tgts_get(obj, DAOS_OBJ_RPC_UPDATE, dkey_hash, tgt_set,
+			      map_ver, false, &obj_auxi->req_tgts);
+	if (rc)
```

Comment:
Hmm, probably tgts_get() should be together with ec_update_encode(), then we can totally separate ec and replication, i.e. probably different ops for ec and replication, for example do something like
oca->oca_split_ios(obj, iod, sals, &tgts, &tgt_cnt);

Reply:
I am not sure if it is good to add the ops to oca, maybe need not.
i'll think more, if change that it could be in following patch.